### PR TITLE
Improving load balancer performance

### PR DESCRIPTION
**Pull request by ahjumma**
TL;DR
Updates to improve load balancer performance at a high load scenario.

Resolves https://github.com/moby/moby/issues/35082

Detailed explanation is provided at https://github.com/moby/moby/issues/35082#issuecomment-565245210

### CODE REVIEW

**Comment by arkodg** [comment]
thanks for running the tests and contributing ! Had a few questions - 

1. should we also set `net/ipv4/vs/expire_quiescent_template` to `1` which should expire persistent connections to the real server with weight 0 (after the backend is down thanks to the `net/ipv4/vs/expire_nodest_conn=1` setting ), but I see there is an open issue https://github.com/kubernetes/kubernetes/issues/81775 so would like to understand the negative implications of setting `net/ipv4/vs/conn_reuse_mode` to `0`
cc: @lbernail 

2. will this setting work for most kernels ?

**Comment by geekdave** [comment]
This fix resolved the same issue for me.  My findings here: https://github.com/moby/moby/issues/35082#issuecomment-585433224

Happy to help in any way to move this forward!

**Comment by geekdave** [comment]
@arkodg do you think this fix is likely to be introduced at this layer?  Trying to understand if we should add scripting to our deployment that adds this after every docker stack deploy:

```
sudo nsenter --net=/var/run/docker/netns/{your_load_balancer} sysctl -w net.ipv4.vs.conn_reuse_mode=0
sudo nsenter --net=/var/run/docker/netns/{your_load_balancer} sysctl -w net.ipv4.vs.expire_nodest_conn=1
```

...or whether we should wait for a fix here.  Any guidance would be much appreciated!

**Comment by ahjumma** [comment]
@geekdave 
That's our current solution to this problem.

```
default_network_id=$(docker network ls | grep "myservice_default" | awk '{print $1}')
sudo nsenter --net=/var/run/docker/netns/lb_${default_network_id:0:9} sysctl -w net.ipv4.vs.conn_reuse_mode=0
sudo nsenter --net=/var/run/docker/netns/lb_${default_network_id:0:9} sysctl -w net.ipv4.vs.expire_nodest_conn=1
```
Once your stack is running, consecutive *docker stack deploy* is unnecessary since your load balancer sandbox will be re-used throughout the lifetime of your docker stack.

**Comment by geekdave** [comment]
Thanks for this, @ahjumma !  We have some tooling that performs an initial stack deploy on fresh machines.  I'm considering adding some scripting that waits in a loop until the `${service}_default` network is created before running your commands.  Does that seem like the right approach?

**Comment by lbernail** [comment]
👋 In kube-proxy we set:
- net.ipv4.vs.conn_reuse_mode=0
- net.ipv4.vs.expire_nodest_conn=1
- net.ipv4.vs.expire_quiescent_template=1

expire_nodest_conn is required to avoid blackholing traffic to existing destinations after removing the RealServer (RST on next packet to backend for TCP, ICMP destination unreachable for UDP)

expire_quiescent_template is great when you set weight to 0 to avoid creating new connections to a real server with weight 0

conn_reuse_mode is great for performances: if you set it to 1, a new connection reusing the same 5 tuples will get its first packet dropped, which means the new connection will wait for 1s to retry.
However, it leads to some problems in kube-proxy because if ports are reused fast enough a realserver always has Active/InActive Connections and is never garbage collected (we currently asynchronously gc backends with weight 0 when they have no connections, but we are looking into ways to forcefully delete the backend after a while to avoid this situation)

**Comment by ahjumma** [comment]
> thanks for running the tests and contributing ! Had a few questions -
> 
> 1. should we also set `net/ipv4/vs/expire_quiescent_template` to `1` which should expire persistent connections to the real server with weight 0 (after the backend is down thanks to the `net/ipv4/vs/expire_nodest_conn=1` setting ), but I see there is an open issue [kubernetes/kubernetes#81775](https://github.com/kubernetes/kubernetes/issues/81775) so would like to understand the negative implications of setting `net/ipv4/vs/conn_reuse_mode` to `0`
>    cc: @lbernail
> 2. will this setting work for most kernels ?

Hi @arkodg.

It does seem **net.ipv4.vs.expire_quiescent_template=1** would be also helpful. 

Regarding your concern raised by possible problematic behaviours of **net.ipv4.vs.conn_reuse_mode=0**, **net.ipv4.vs.expire_nodest_conn=1** is what allows by-passing the problem.
This has been acknowledged in [kubernetes/kubernetes/issues/81775#issuecomment-542542167](https://github.com/kubernetes/kubernetes/issues/81775#issuecomment-542542167) you have referenced .

For the second question, what's the earliest linux kernel that Docker Swarm needs to support?

**Comment by arkodg** [comment]
thanks @lbernail @ahjumma 
can we also include  `net.ipv4.vs.expire_quiescent_template=1`

**Comment by ahjumma** [comment]
Hi @arkodg ,

According to @lbernail's previous [finding](https://github.com/kubernetes/kubernetes/pull/71834#issuecomment-445175643), kernel 2.6 is supported.

File: osl/namespace_linux.go [hunk 1]
```diff
@@ -38,9 +38,15 @@ var (
 	gpmChan            = make(chan chan struct{})
 	prefix             = defaultPrefix
 	loadBalancerConfig = map[string]*kernel.OSValue{
+		// disables any special handling on port reuse of existing IPVS connection table entries
+		// more info: https://github.com/torvalds/linux/blob/master/Documentation/networking/ipvs-sysctl.txt#L25:1
+		"net.ipv4.vs.conn_reuse_mode": {Value: "0", CheckFn: nil},
 		// expires connection from the IPVS connection table when the backend is not available
 		// more info: https://github.com/torvalds/linux/blob/master/Documentation/networking/ipvs-sysctl.txt#L126:1
 		"net.ipv4.vs.expire_nodest_conn": {Value: "1", CheckFn: nil},
+		// expires persistent connections to destination servers with weights set to 0
```
Comment (arkodg) [inline osl/namespace_linux.go]: LGTM

**Comment by arkodg** [comment]
can you ptal as well @lbernail

**Comment by selansen** [comment]
will look into it soon.

**Comment by cuongnv4sapo** [comment]
I did the same steps but the ingress ld still low. Did I missed any steps

sudo nsenter --net=/var/run/docker/netns/lb_o7wjqxluj sysctl  net.ipv4.vs.conn_reuse_mode
sudo nsenter --net=/var/run/docker/netns/lb_o7wjqxluj sysctl  net.ipv4.vs.expire_nodest_conn
sudo nsenter --net=/var/run/docker/netns/lb_o7wjqxluj sysctl  net.ipv4.vs.expire_quiescent_template

**Pull request merged**